eq_cir_converter_service/utils/helper_utils.py

```diff
@@ -0,0 +1,209 @@
+"""Utility functions for text processing."""
+
+import re
+from collections import Counter
+from collections.abc import Sequence
+from copy import deepcopy
+from typing import Any
+
+from eq_cir_converter_service.config.logging_config import logging
+
+logger = logging.getLogger(__name__)
+
+# --- Regular Expressions ---
+# Compiled regular expressions for HTML tag processing
+REGEX_B_OPEN = re.compile(r"<\s*b\s*>", flags=re.IGNORECASE)
+REGEX_B_CLOSE = re.compile(r"<\s*/\s*b\s*>", flags=re.IGNORECASE)
+REGEX_P_TAGS = re.compile(r"</?p>", flags=re.IGNORECASE)
+REGEX_BR_TAGS = re.compile(r"</?br>", flags=re.IGNORECASE)
+REGEX_PARAGRAPH_SPLIT = re.compile(r"<p>(.*?)</p>", flags=re.IGNORECASE | re.DOTALL)
+REGEX_PLACEHOLDER = re.compile(r"\{(.*?)\}", flags=re.IGNORECASE)
+
+
+# --- HTML Tag Processing ---
+def remove_and_replace_tags(text: str) -> str:
+    """Cleans HTML tags from the text, replacing <b> with <strong> and removing <br> and <p> tags.
+
+    :param text: The input text containing HTML tags.
+    :return: The cleaned text with HTML tags removed or replaced.
+    """
+    return REGEX_P_TAGS.sub(
+        "",
+        REGEX_BR_TAGS.sub(
+            "",
+            REGEX_B_CLOSE.sub(
+                "</strong>",
+                REGEX_B_OPEN.sub("<strong>", text),
+            ),
+        ),
+    ).strip()
+
+
+# --- Paragraph Extraction ---
+def split_paragraphs_into_list(paragraphs_string: str) -> list[str]:
+    """Extracts paragraphs from string, returning a list of cleaned paragraph strings.
+
+    :param paragraphs_string: The string to extract paragraphs from.
+    :return: A list of cleaned paragraphs.
+    """
+    paragraphs = REGEX_PARAGRAPH_SPLIT.findall(paragraphs_string)
+    # Remove empty items from the list (empties being side effects of regex)
+    return [paragraph.strip() for paragraph in paragraphs if paragraph.strip()]
+
+
+# --- Placeholder Extraction ---
+def extract_placeholder_names_from_text_field(text: str) -> list[str]:
+    """Extracts placeholders from the text, returning a list of placeholder names.
+
+    :param text: The input text containing placeholders.
+    :return: A list of placeholder names.
+    """
+    return REGEX_PLACEHOLDER.findall(text)
+
+
+# --- Handle "text" with "placeholders" ---
+def split_paragraphs_with_placeholders(
```

Comment:
As discussed, I think we can make the codebase clearer here and improve readability by bumping some of the inline comments up to the docstring, and renaming some of the variables e.g `paragraphs` instead of `paraghraphs_list` so it is slightly less verbose. Also would be good to add an example transform in the docstring for the placeholder for clarity.